modelhub/modelhub/models/funnel_discovery.py

```diff
@@ -0,0 +1,329 @@
+        )
+
+        window = nav_df.groupby(by=nav_df.index_columns).window()
+        root_step_series = window[flattened_lc.name]
+
+        all_step_series = {}
+        for step in range(1, steps + 1):
+            step_series_name = f'{flattened_lc.name}_step_{step}'
+            if step == 1:
+                next_step = root_step_series.copy_override(group_by=None)
+            else:
+                next_step = root_step_series.window_lag(offset=step - 1)
+
+            all_step_series[step_series_name] = (
+                next_step.copy_override(name=step_series_name)
+            )
+
+        result = nav_df.copy_override(
+            base_node=window.base_node,
+            series={
+                **all_step_series,
+                offset_lc.name: offset_lc.copy_override(base_node=window.base_node),
+            }
+        )
+        result = result.materialize(node_name='step_extraction')
+
+        # limit rows where initial step is between the requested steps
+        result = result[result[offset_lc.name] < steps]
+
+        # re-order rows
+        result = result.sort_values(by=result.index_columns + [offset_lc.name])
+        result = result.drop(columns=[offset_lc.name])
+        return result
+
+    def filter_navigation_paths_to_conversion(self, steps_df, conversion_events_df,
+                                              conversion_location_column='feature_nice_name'
+                                              ) -> bach.DataFrame:
+        """
+        Filter steps dataframe rows to first conversion location.
+        For each row of steps_df dataframe set to None the step values
+         after encountering the first conversion location step.
+
+        Corner case: in case the first location stack is conversion one we ignore it
+        and proceed to the next step.
+
+        :param steps_df: dataframe which one gets from `get_navigation_paths` method output.
+            This the dataframe that we're going to filter.
+        :param conversion_events_df: dataframe where for each `conversion_location_column`
+            value we have info if there was a conversion event. We will merge this dataframe
+            with 'melted' steps_df in order to have `is_conversion_event` column for steps_df.
+        :param conversion_location_column: need this column in the dataframes merge in order
+            to get converted steps df.
+
+        :returns: bach DataFrame, filtered each row of steps_df to the
+        """
+
+        conversion_events_df_columns = [conversion_location_column, 'is_conversion_event']
+        if not all(col in conversion_events_df.data_columns
+                   for col in conversion_events_df_columns):
+            raise ValueError(f'The {", ".join(conversion_events_df_columns)} columns should be in the main dataframe')
+
+        _steps_df = steps_df.copy()
+        _steps_df['_index'] = steps_df.groupby().window().location_stack_step_1.window_row_number()
```

Comment:
I will be carefull with this line, what if the location stack event has actually a different name?
```suggestion
        _steps_df['_index'] = steps_df.groupby().window().location_stack_step_1.window_row_number()
```